openshift/dynamic/client.py

```diff
@@ -0,0 +1,394 @@
+
+        if is_openshift:
+            groups['oapi'] = { '': {
+                'v1': self.get_resources_for_api_version('oapi', '', 'v1', True)
+            }}
+        return groups
+
+    def parse_api_groups(self):
+        ""
+        prefix = 'apis'
+        groups_response = self.request('GET', '/{}'.format(prefix))['groups']
+
+        groups = self.default_groups()
+        groups[prefix] = {}
+
+        for group in groups_response:
+            new_group = {}
+            for version_raw in group['versions']:
+                version = version_raw['version']
+                preferred = version_raw == group['preferredVersion']
+                new_group[version] = self.get_resources_for_api_version(prefix, group['name'], version, preferred)
+            groups[prefix][group['name']] = new_group
+        return groups
+
+    def get_resources_for_api_version(self, prefix, group, version, preferred):
+        """ returns a dictionary of resources associated with provided groupVersion"""
+
+        resources = {}
+
+        path = '/'.join(filter(None, [prefix, group, version]))
+        resources_response = self.request('GET', path)['resources']
+
+        # TODO: Filter out subresources for now
+        resources_raw = filter(lambda resource: '/' not in resource['name'], resources_response)
+
+        for resource in resources_raw:
+            resources[resource['kind']] = Resource(
+                prefix=prefix,
+                group=group,
+                api_version=version,
+                client=self,
+                preferred=preferred,
+                **resource
+            )
+        return resources
+
+    def list(self, resource, namespace=None, label_selector=None, field_selector=None):
+        path_params = {}
+        if resource.namespaced and namespace:
+            resource_path = resource.urls['namespaced_base']
+            path_params['namespace'] = namespace
+        else:
+            resource_path = resource.urls['base']
+        return ResourceInstance(resource, self.request('get', resource_path, path_params=path_params, label_selector=label_selector, field_selector=field_selector))
+
+    def get(self, resource, name=None, namespace=None, label_selector=None, field_selector=None):
+        if name is None:
+            return self.list(resource, namespace=namespace, label_selector=label_selector, field_selector=field_selector)
+        path_params = {'name': name}
+        if resource.namespaced and namespace:
+            resource_path = resource.urls['namespaced_full']
+            path_params['namespace'] = namespace
+        else:
+            resource_path = resource.urls['full']
+        return ResourceInstance(resource, self.request('get', resource_path, path_params=path_params, label_selector=label_selector, field_selector=field_selector))
+
+    def create(self, resource, body, namespace=None):
+        path_params = {}
+        if resource.namespaced and namespace:
```

Comment:
I see a lot of duplication with these `if` statements. I notice some result in grabbing `namespaced_full` of `namespaced_base` but maybe this could be simplified?

Reply:
added a path method to resource type, that look better?